File: app/src/main/java/br/com/jonathanzanella/myexpenses/chargeable/ListChargeableActivity.java
package br.com.jonathanzanella.myexpenses.chargeable;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.account.AccountAdapter;
import br.com.jonathanzanella.myexpenses.account.AccountAdapterCallback;
import br.com.jonathanzanella.myexpenses.card.Card;
import br.com.jonathanzanella.myexpenses.card.CardAdapter;
import br.com.jonathanzanella.myexpenses.card.CardAdapterCallback;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.Bind;

public class ListChargeableActivity extends BaseActivity implements AccountAdapterCallback, CardAdapterCallback {
	public static final String KEY_CHARGEABLE_SELECTED_UUID = "KeyChargeableSelectUuid";
	public static final String KEY_CHARGEABLE_SELECTED_TYPE = "KeyChargeableSelectType";

	@Bind(R.id.act_chargeable_list_accounts)
	RecyclerView accounts;
	@Bind(R.id.act_chargeable_list_cards)
	RecyclerView cards;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_list_chargeable);
		setTitle(R.string.select_chargeable_title);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);

		initAccounts();
		initCreditCards();
	}

	private void initAccounts() {
		AccountAdapter adapter = new AccountAdapter();
		adapter.setCallback(this);
		adapter.setFormat(AccountAdapter.Format.LIST);

		accounts.setAdapter(adapter);
		accounts.setHasFixedSize(true);
		accounts.setLayoutManager(new GridLayoutManager(this, 2));
		accounts.setItemAnimator(new DefaultItemAnimator());
	}

	private void initCreditCards() {
		final CardAdapter adapter = new CardAdapter();
		adapter.setCallback(this);
		new AsyncTask<Void, Void, Void>() {

			@Override
			protected Void doInBackground(Void... voids) {
				adapter.loadData();
				return null;
			}

			@Override
			protected void onPostExecute(Void aVoid) {
				super.onPostExecute(aVoid);
				cards.setAdapter(adapter);
				cards.setHasFixedSize(true);
				cards.setLayoutManager(new GridLayoutManager(ListChargeableActivity.this, 2));
				cards.setItemAnimator(new DefaultItemAnimator());
			}
		}.execute();
	}

	@Override
	public void onAccountSelected(Account account) {
		Intent i = new Intent();
		i.putExtra(KEY_CHARGEABLE_SELECTED_UUID, account.getUuid());
		i.putExtra(KEY_CHARGEABLE_SELECTED_TYPE, account.getChargeableType());
		setResult(RESULT_OK, i);
		finish();
	}

	@Override
	public void onCard(Card card) {
		Intent i = new Intent();
		i.putExtra(KEY_CHARGEABLE_SELECTED_UUID, card.getUuid());
		i.putExtra(KEY_CHARGEABLE_SELECTED_TYPE, card.getChargeableType());
		setResult(RESULT_OK, i);
		finish();
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/chargeable/ListChargeableActivity.kt
package br.com.jonathanzanella.myexpenses.chargeable

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.account.AccountAdapter
import br.com.jonathanzanella.myexpenses.account.AccountAdapterCallback
import br.com.jonathanzanella.myexpenses.card.Card
import br.com.jonathanzanella.myexpenses.card.CardAdapter
import br.com.jonathanzanella.myexpenses.card.CardAdapterCallback
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.jetbrains.anko.*
import javax.inject.Inject

class ListChargeableActivity : AppCompatActivity(), AccountAdapterCallback, CardAdapterCallback {
    @Inject
    lateinit var adapter: AccountAdapter
    private val ui = ListChargeableActivityUi()

    init {
        App.getAppComponent().inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        ui.toolbar.title = getString(R.string.select_chargeable_title)
        ui.toolbar.setup(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        initAccounts()
        initCreditCards()
    }

    override fun onDestroy() {
        super.onDestroy()
        adapter.onDestroy()
    }

    private fun initAccounts() {
        adapter.setCallback(this)
        adapter.setFormat(AccountAdapter.Format.LIST)

        ui.accounts.adapter = adapter
        ui.accounts.setHasFixedSize(true)
        ui.accounts.layoutManager = GridLayoutManager(this, 2)
        ui.accounts.itemAnimator = DefaultItemAnimator()
    }

    private fun initCreditCards() {
        val cardAdapter = CardAdapter()
        cardAdapter.setCallback(this)

        doAsync {
            cardAdapter.loadData()

            uiThread {
                ui.cards.apply {
                    adapter = cardAdapter
                    setHasFixedSize(true)
                    layoutManager = GridLayoutManager(this@ListChargeableActivity, 2)
                    itemAnimator = DefaultItemAnimator()
                }
            }
        }
    }

    override fun onAccountSelected(account: Account) {
        val i = Intent()
        i.putExtra(KEY_CHARGEABLE_SELECTED_UUID, account.uuid)
        i.putExtra(KEY_CHARGEABLE_SELECTED_TYPE, account.chargeableType)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    override fun onCard(card: Card) {
        val i = Intent()
        i.putExtra(KEY_CHARGEABLE_SELECTED_UUID, card.uuid)
        i.putExtra(KEY_CHARGEABLE_SELECTED_TYPE, card.chargeableType)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    companion object {
        val KEY_CHARGEABLE_SELECTED_UUID = "KeyChargeableSelectUuid"
        val KEY_CHARGEABLE_SELECTED_TYPE = "KeyChargeableSelectType"
    }
}

private class ListChargeableActivityUi : AnkoComponent<ListChargeableActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var accounts: RecyclerView
    lateinit var cards: RecyclerView

    override fun createView(ui: AnkoContext<ListChargeableActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}

            headerTitle { text = resources.getString(R.string.accounts) }
            accounts = recyclerView { id = R.id.act_chargeable_list_accounts }.lparams(height = 0, weight = 1f)

            headerTitle { text = resources.getString(R.string.cards) }
                    .lparams { topMargin = resources.getDimensionPixelSize(R.dimen.default_spacing) }
            cards = recyclerView { id = R.id.act_chargeable_list_cards }.lparams(height = 0, weight = 1f)
        }
    }.applyRecursively(::applyTemplateViewStyles)
}
